src/Components/Web/src/Virtualization/Virtualize.cs

```diff
@@ -0,0 +1,304 @@
+                    builder.AddContent(renderIndex, _itemTemplate, item);
+                    renderIndex++;
+                }
+
+                builder.CloseRegion();
+            }
+
+            builder.OpenRegion(5);
+
+            // Render the placeholders after the loaded items.
+            for (; renderIndex < lastItemIndex; renderIndex++)
+            {
+                builder.AddContent(renderIndex, _placeholder, new PlaceholderContext(renderIndex));
+            }
+
+            builder.CloseRegion();
+
+            var itemsAfter = Math.Max(0, _itemCount - _visibleItemCapacity - _itemsBefore);
+
+            builder.OpenElement(6, "div");
+            builder.AddAttribute(7, "style", GetSpacerStyle(itemsAfter));
+            builder.AddElementReferenceCapture(8, elementReference => _spacerAfter = elementReference);
+
+            builder.CloseElement();
+        }
+
+        private string GetSpacerStyle(int itemsInSpacer)
+            => $"height: {itemsInSpacer * ItemSize}px;";
+
+        void IVirtualizeJsCallbacks.OnBeforeSpacerVisible(float spacerSize, float containerSize)
+        {
+            CalcualteItemDistribution(spacerSize, containerSize, out var itemsBefore, out var visibleItemCapacity);
+
+            UpdateItemDistribution(itemsBefore, visibleItemCapacity);
+        }
+
+        void IVirtualizeJsCallbacks.OnAfterSpacerVisible(float spacerSize, float containerSize)
+        {
+            CalcualteItemDistribution(spacerSize, containerSize, out var itemsAfter, out var visibleItemCapacity);
+
+            var itemsBefore = Math.Max(0, _itemCount - itemsAfter - visibleItemCapacity);
+
+            UpdateItemDistribution(itemsBefore, visibleItemCapacity);
+        }
+
+        private void CalcualteItemDistribution(float spacerSize, float containerSize, out int itemsInSpacer, out int visibleItemCapacity)
+        {
+            itemsInSpacer = Math.Max(0, (int)Math.Floor(spacerSize / ItemSize) - 1);
+            visibleItemCapacity = (int)Math.Ceiling(containerSize / ItemSize) + 2;
+        }
+
+        private void UpdateItemDistribution(int itemsBefore, int visibleItemCapacity)
+        {
+            if (itemsBefore != _itemsBefore || visibleItemCapacity != _visibleItemCapacity)
+            {
+                _itemsBefore = itemsBefore;
+                _visibleItemCapacity = visibleItemCapacity;
+                _refreshTask = RefreshDataAsync();
+
+                if (!_refreshTask.IsCompleted)
+                {
+                    StateHasChanged();
+                }
+            }
+        }
+
+        private async Task RefreshDataAsync()
+        {
+            _refreshCts?.Cancel();
+            _refreshCts = new CancellationTokenSource();
+
+            var cancellationToken = _refreshCts.Token;
+            var request = new ItemsProviderRequest(_itemsBefore, _visibleItemCapacity, cancellationToken);
+
+            try
+            {
+                var result = await _itemsProvider(request);
```

Comment:
Should we clear `_refreshException` prior to invoking this?

Reply:
How about clearing it right before it's thrown in `BuildRenderTree`? So at the start of `BuildRenderTree`, something like:
```
if (_refreshException != null)
{
    var oldRefreshException = _refreshException;
    _refreshException = null;

    throw oldRefreshException;
}
```